Add definition_of_ready.md (#11)
PilzDE/.github: definition_of_ready.md

# Definition of ready (DoR)
- Each user story should fulfill the INVEST criteria (see e.g. [Wikipedia](https://en.wikipedia.org/wiki/INVEST_(mnemonic)), [srum-in-der-praxis.de](https://scrum-in-der-praxis.de/glossary/invest))
  - See https://www.knowledgehut.com/tutorials/scrum-tutorial/definition-of-ready
    - Independent (of all others)
    - Negotiable (not a contract for a specific feature)
    - Valuable (a vertical slice)
    - Estimable (to a decent approximation)
    - Small (enough to fit in a single iteration)
    - Testable (even if the test doesn't exist yet)

> ## In detail

Reviewer: In relation to the INVEST criteria, the following is an addition, I think
```suggestion
## Further details
```

- The user story is Well-defined:

>   - The user story should be written in the user story format, see for example [Wikipedia](https://en.wikipedia.org/wiki/User_story#Principle).

Reviewer: ```suggestion
  - The user story should be written in the user story format, see for example [Wikipedia](https://en.wikipedia.org/wiki/User_story#Principle).
  - e.g. `As a <role> I can <capability>, so that <receive benefit>`
```

- Acceptance criteria are defined (see point 7 from [dzone.com](https://dzone.com/articles/10-tips-for-writing-good-user-stories))

>   - The conditions are described which have to be fulfilled so that the story is done.

Reviewer: ```suggestion
  - The conditions that have to be fulfilled such that the story is done are described. The description should allow any member of the dev team to decide whether the story is done without further communication with PO or stakeholders.
```

- Diagrams are used to support the understanding (see point 10 from [dzone.com](https://dzone.com/articles/10-tips-for-writing-good-user-stories))
- No implicit knowledge
  - All information regarding a User-story should be written down in easy to understand sentences. Reason: It only takes a few minutes to write valuable information down but it takes ages to fix implementations based on incorrectly understood/missing information.
   - Whenever new knowledge is gained regarding the User-story, it should be written down in the User-story, to promote information exchange.
 - "Dependencies are identified and no external dependencies would block the PBI from being completed." (Source: [informit.com](https://www.informit.com/articles/article.aspx?p=1928232&seqNum=5))
 - If external documents (e.g. specs) are needed for the story they are collected and provided with the story.
 - If external help is needed the feasibility of getting this help is ensured. (Maybe even dates are set)

> 

Reviewer: Kann man das so sagen?
```suggestion
  - External dependencies may be present at the start of the PBI but it must be foreseeable that they are solved by the end of the work

```

    - Negotiable (not a contract for a specific feature)
    - Valuable (a vertical slice)
    - Estimable (to a decent approximation)
    - Small (enough to fit in a single iteration)
    - Testable (even if the test doesn't exist yet)

## In detail
- The user story is Well-defined:
  - The user story should be written in the user story format, see for example [Wikipedia](https://en.wikipedia.org/wiki/User_story#Principle).
- Acceptance criteria are defined (see point 7 from [dzone.com](https://dzone.com/articles/10-tips-for-writing-good-user-stories))
  - The conditions are described which have to be fulfilled so that the story is done.
  - The team knows how to "demonstrate" the user story to the user. In other words, the team knows the value for the user/stakeholder.
  - "Without such criteria developers may not really understand the scope of the work or how it will be tested and validated." (Source: [scrum.org](https://www.scrum.org/resources/blog/walking-through-definition-ready))
- Diagrams are used to support the understanding (see point 10 from [dzone.com](https://dzone.com/articles/10-tips-for-writing-good-user-stories))
- No implicit knowledge

>   - All information regarding a User-story should be written down in easy to understand sentences. Reason: It only takes a few minutes to write valuable information down but it takes ages to fix implementations based on incorrectly understood/missing information.

Reviewer 2: I would say: If no one has open questions, the information is complete.

>    - Whenever new knowledge is gained regarding the User-story, it should be written down in the User-story, to promote information exchange.

Reviewer 2: What do you think about: https://github.com/PilzDE/.github/pull/11#discussion_r442754091

# Definition of ready (DoR)
- Each user story should fulfill the INVEST criteria (see e.g. [Wikipedia](https://en.wikipedia.org/wiki/INVEST_(mnemonic)), [scrum-in-der-praxis.de](https://scrum-in-der-praxis.de/glossary/invest))
  - See https://www.knowledgehut.com/tutorials/scrum-tutorial/definition-of-ready
    - Independent (of all others)
    - Negotiable (not a contract for a specific feature)
    - Valuable (a vertical slice)
    - Estimable (to a decent approximation)
    - Small (enough to fit in a single iteration)
    - Testable (even if the test doesn't exist yet)

## Further requirements

> - The user story is Well-defined:

Reviewer 3: What about non-user-stories? Issues? Tutorials?

Reviewer 4: Imo everything can be written as a user-story. (whether that makes sense is another point of discussion though)

  - The user story should be written in the user story format, see for example [Wikipedia](https://en.wikipedia.org/wiki/User_story#Principle).
  - e.g. `As a <role> I can <capability>, so that <receive benefit>`
- Acceptance criteria are defined (see point 7 from [dzone.com](https://dzone.com/articles/10-tips-for-writing-good-user-stories))

>   - The conditions that have to be fulfilled such that the story is done are described. The description should allow any member of the dev team to decide whether the story is done.

Reviewer 2: ```suggestion
  - The description states clearly which conditions have to be fulfilled such that the story is done.
```

Reviewer 2: What do you think? I omitted the "team"-part here, because I think the whole story should be easy to understand, which I included in the succestion below.

  - The team knows how to "demonstrate" the user story to the user. In other words, the team knows the value for the user/stakeholder.
  - "Without such criteria developers may not really understand the scope of the work or how it will be tested and validated." (Source: [scrum.org](https://www.scrum.org/resources/blog/walking-through-definition-ready))
- Diagrams are used to support the understanding (see point 10 from [dzone.com](https://dzone.com/articles/10-tips-for-writing-good-user-stories))
- No implicit knowledge
  - If possible, all information that the team requires to understand a User-story should be written down.

>   - Whenever new knowledge is gained regarding the User-story, its description should be adjusted.

Reviewer 2: ```suggestion
  - The whole team tries to keep the User-story and its description up-to-date.
```

> - No implicit knowledge
>   - All information regarding a User-story should be written down in easy to understand sentences. Reason: It only takes a few minutes to write valuable information down but it takes ages to fix implementations based on incorrectly understood/missing information.

Reviewer 2: ```suggestion
- Easy to understand
  - If possible, all information that the team requires to understand a User-story should be written down.
```

Reviewer: I would prefer to keep `- No implicit knowledge`

>    - Whenever new knowledge is gained regarding the User-story, it should be written down in the User-story, to promote information exchange.

Reviewer 2: ```suggestion
   - Whenever new knowledge is gained regarding the User-story, its description should be adjusted.
```

> - No implicit knowledge
>   - All information regarding a User-story should be written down in easy to understand sentences. Reason: It only takes a few minutes to write valuable information down but it takes ages to fix implementations based on incorrectly understood/missing information.
>    - Whenever new knowledge is gained regarding the User-story, it should be written down in the User-story, to promote information exchange.

Reviewer 3: ```suggestion
- Easy to understand: The team writes down or attaches additional information, that it requires to understand a user-story.
```

Reviewer 2: I would not write it like that, because I think ideally the final description of the user-story is worked out together.
